### PR TITLE
Fixes #20377 - Add features for async-ssh callback

### PR DESCRIPTION
**Pull request by adamruzicka**

### CODE REVIEW

File: lib/smart_proxy_dynflow_core/settings.rb [hunk 1]
```diff
@@ -40,7 +40,8 @@ class Settings < OpenStruct
         :plugins => {},
         :pid_file => '/var/run/foreman-proxy/smart_proxy_dynflow_core.pid',
         :daemonize => false,
-        :loaded => false
+        :loaded => false,
+        :tokens => []
```
Comment (iNecas) [inline lib/smart_proxy_dynflow_core/settings.rb]: Is this being used anywhere now?

**Comment by iNecas** [comment]
Please file different issue for each of the PRs, as I try to keep the release attribute on the redmine to match the gem versions. We can have the interlinking to the issues in the redmine then. It might help us not getting lost with the multi-prs featues

File: lib/smart_proxy_dynflow_core/settings.rb (hunk 1 above)
Comment (adamruzicka) [inline lib/smart_proxy_dynflow_core/settings.rb]: No, will remove

**Comment by iNecas** [comment]
foreman-tasks-core 0.2.1 was released. Please update the dependencies. Also, similarly to foreman-tasks PR, some test coverage would be good to have before merging

**Comment by iNecas** [comment]
While doing so, it will be also needed to fix the tests, that are failing at the moment :/

**Comment by adamruzicka** [comment]
Fixed the broken tests and added a bunch of new ones

**Comment by adamruzicka** [comment]
Found and issue, do not merge

**Comment by adamruzicka** [comment]
Issue fixed, this should be ready to go now

File: lib/smart_proxy_dynflow_core/api.rb [hunk 2]
```diff
@@ -7,13 +7,14 @@ class Api < ::Sinatra::Base
...
-      authorize_with_ssl_client
+      authorize_with_token || authorize_with_ssl_client
       content_type :json
     end
 
     post "/tasks/?" do
       params = MultiJson.load(request.body.read)
-      trigger_task(::Dynflow::Utils.constantize(params['action_name']), params['action_input']).to_json
+      trigger_task(::Dynflow::Utils.constantize(params['action_name']),
+                   params['action_input'].merge(:callback_host => request.env['HTTP_X_FORWARDED_FOR'])).to_json
```
Comment (adamruzicka) [inline lib/smart_proxy_dynflow_core/api.rb]: Note: `X-Forwarded-For` won't be set if dynflow_core is not a standalone process, will have to fallback to `HTTP_HOST` in that case

File: smart_proxy_dynflow_core.gemspec [hunk 3]
```diff
@@ -30,7 +30,7 @@ Gem::Specification.new do |gem|
   gem.add_development_dependency('rack-test', '~> 0')
   gem.add_development_dependency('rubocop', '0.32.1')
 
-  gem.add_runtime_dependency('foreman-tasks-core', '>= 0.1')
+  gem.add_runtime_dependency('foreman-tasks-core', '>= 0.2.1')
```
Comment (iNecas) [inline smart_proxy_dynflow_core.gemspec]: I've backported the requirements to `core-0.1.5` so that we can still release this with foreman 1.15

**Comment by iNecas** [comment]
Needs rebase, tested in production and works well. Let's merge after jenkins turns green

File: smart_proxy_dynflow_core.gemspec [hunk 4]
```diff
@@ -30,8 +30,8 @@ Gem::Specification.new do |gem|
   gem.add_development_dependency('rack-test', '~> 0')
   gem.add_development_dependency('rubocop', '0.32.1')
 
-  gem.add_runtime_dependency('foreman-tasks-core', '>= 0.1')
   gem.add_runtime_dependency('dynflow', "~> 0.8.27")
+  gem.add_runtime_dependency('foreman-tasks-core', '>= 0.2.1')
```
Comment (iNecas) [inline smart_proxy_dynflow_core.gemspec]: I believe this should be >= '0.1.5', as we backported the required changes to tasks `0.9.x` branch to keep it compatilbe with Foreman 1.15

Comment (adamruzicka) [reply smart_proxy_dynflow_core.gemspec]: Right, missed that during rebase

**Comment by iNecas** [comment]
Master is red due to intermittent issues: I'm trying to address it here https://github.com/theforeman/smart_proxy_dynflow/pull/38

**Comment by adamruzicka** [comment]
We're green here as well

**Comment by iNecas** [comment]
Thanks @adamruzicka

**Pull request merged**